test/foundry/src/concrete/vault/OffchainAssetReceiptVault.tiers.t.sol

```diff
@@ -0,0 +1,218 @@
+// SPDX-License-Identifier: CAL
+pragma solidity =0.8.25;
+
+import {
+    UnauthorizedSenderTier,
+    OffchainAssetReceiptVault
+} from "../../../../../contracts/concrete/vault/OffchainAssetReceiptVault.sol";
+import {OffchainAssetReceiptVaultTest, Vm} from "test/foundry/abstract/OffchainAssetReceiptVaultTest.sol";
+import {LibOffchainAssetVaultCreator} from "test/foundry/lib/LibOffchainAssetVaultCreator.sol";
+import {ITierV2} from "rain.tier.interface/interface/ITierV2.sol";
+
+contract TiersTest is OffchainAssetReceiptVaultTest {
+    event SetERC20Tier(address sender, address tier, uint256 minimumTier, uint256[] context, bytes data);
+    event SetERC1155Tier(address sender, address tier, uint256 minimumTier, uint256[] context, bytes data);
+
+    /// Test setERC20Tier event
+    function testSetERC20Tier(
+        uint256 fuzzedKeyAlice,
+        uint256 fuzzedKeyBob,
+        string memory assetName,
+        string memory assetSymbol,
+        bytes memory fuzzedData,
+        uint8 fuzzedMinTier,
+        uint256[] memory fuzzedContext,
+        address tier
+    ) external {
+        // Ensure the fuzzed key is within the valid range for secp256k1
+        address alice = vm.addr((fuzzedKeyAlice % (SECP256K1_ORDER - 1)) + 1);
+        address bob = vm.addr((fuzzedKeyBob % (SECP256K1_ORDER - 1)) + 1);
+        vm.assume(alice != bob);
+
+        fuzzedMinTier = uint8(bound(fuzzedMinTier, uint256(1), uint256(8)));
+
+        // Create the vault
+        OffchainAssetReceiptVault vault = createVault(alice, assetName, assetSymbol);
+
+        // Prank as Alice to grant roles
+        vm.startPrank(alice);
+
+        // Grant the necessary role
+        vault.grantRole(vault.ERC20TIERER(), bob);
+        vm.stopPrank();
+
+        // Prank as Bob
+        vm.startPrank(bob);
+
+        // Emit the expected event
+        vm.expectEmit(true, true, true, true);
+        emit SetERC20Tier(bob, address(tier), fuzzedMinTier, fuzzedContext, fuzzedData);
+
+        // Call the function that should emit the event
+        vault.setERC20Tier(address(tier), fuzzedMinTier, fuzzedContext, fuzzedData);
```

Comment:
so this emits an event but do we have a test that changing the tier contract actually controls how tokens can be moved?